### PR TITLE
Dataframe Collection (User Collections)

### LINKED ISSUE #162: Add support for user-created collections in PyDough
The goal of this issue is to add support for a new type of collection that, instead of corresponding to a collection in the metadata, corresponds to a collection defined by the user, such as:
- Using a PyDough API to parametrically generate a table (e.g. a date range or numerical range)
- Converting a user-providing Pandas DataFrame, or a constant list of values, to a collection (either writing to the database connection, perhaps as a temporary table, or treating as a `VALUES` clause)
- Referring to a table in the database that is not part of the metadata

This issue requires designing the approach for such collections in unqualified nodes, qualification, QDAG nodes, hybrid conversion, hybrid nodes, relational conversion, relational nodes, and SQLGlot conversion, and implementing them for at least one of the choices mentioned above, as well as picking the user syntax for it. The other forms of user-created collections will be followup issues.

Example of what this could look like, with a range of integers becoming a collection:

```py
# Creates a collection `sizes` with a single property `part_size` whose values are the
# integers from 1 (inclusive) to 100 (exclusive), skipping by 5s, then for each size value,
# counts how many turquoise parts have that size.
sizes = pydough.range_collection(1, 100, 5, name='sizes', column=`part_size`)
turquoise_parts = parts.WHERE(CONTAINS(name, 'turquoise'))
result = sizes.CALCULATE(part_size).CALCULATE(part_size, n_parts=COUNT(CROSS(turquoise_parts).WHERE(size == part_size)))
```
In this example, `sizes` could be translated as a VALUES clause, or in Snowflake as a `GENERATOR` table function.

Example of what this could look like, with a pandas DataFrame becoming a collection:

```py
# Creates a collection named `months92` from the DataFrame with columns
# `m` and `d`, with 1 row for each start-of-month date in 1992. Then, for each
# month, count how many urgent orders were placed that month and the change
# from the previous month.
df = pd.DataFrame({
  "m": ["JAN", "FEB", "MAR", "APR", "MAY", "JUN", "JUL", "AUG", "SEP", "OCT", "NOV", "DEC"],
  "d": pd.date_range("1992", "1993", freq="MS", inclusive="left")
})
months = pydough.dataframe_collection(df, name="months92")
urgent_orders = orders.WHERE(order_priority == "1-URGENT")
result = months.CALCULATE(m, d).CALCULATE(n_orders=COUNT(CROSS(urgent_orders).WHERE(d == DATETIME(order_date, "start of month")))).CALCULATE(month=m, n_orders=n_orders, change_from_prev_month=n_orders-PREV(n_orders, by=d.ASC()))
```
In this example, `months` could be translated as a VALUES clause.

The spec for how this should be implemented:
- Each collection-creating API (`dataframe_collection`, `range_collection` etc.) should return a new unqualified node `UnqualifiedGeneratedCollection` which contains an instance of a `PyDoughUserGeneratedCollection` (created by the API).
- There is an abstract base class for `PyDoughUserGeneratedCollection`, with concrete implementation classes for the integer range, DataFrame, and other future implementations, each with some common APIs (e.g. how to stringify, whether it is guaranteed to be non-empty, and what column names/types it has) as the `PyDoughUserGeneratedCollection` instance is daisy-chained throughout the codebase (like how metadata is passed around between IRs)
- There should be a new QDAG node for `GeneratedCollection` (subclass of `ChildAccess`), also containing a `PyDoughUserGeneratedCollection` instance. The qualification of `UnqualifiedGeneratedCollection` to `GeneratedCollection` should be simple by just unwrapping the `PyDoughUserGeneratedCollection` instance from inside the unqualified node and wrapping it in the QDAG node.
- There should be a new hybrid operation node for `HybridGeneratedCollection` (works similar to `HybridCollectionAccess`), also containing a `PyDoughUserGeneratedCollection` instance. The conversion from `GeneratedCollection` to `HybridGeneratedCollection` should also be simple by doing a similar unwrapping & re-wrapping of the underlying `PyDoughUserGeneratedCollection`. Any hybrid utility logic that cases on `HybridCollectionAccess` should probably also case on `HybridGeneratedCollection` (e.g. `always_exists`)
- There should be a new relational node for `GeneratedTable` which works like `Scan` except it takes in the `PyDoughUserGeneratedCollection` (same unwrapping & re-wrapping)
- During SQLGlot conversion, convert the `GeneratedTable` to SQL code based on which underlying `PyDoughUserGeneratedCollection` it uses (& possibly based on the dialect -> can add more stuff to the binding infrastructure)

### PR DESCRIPTION
**Pull request by john-sanchez31**
Resolves #162

### CODE REVIEW

File: documentation/dsl.md [hunk 1]
```diff
@@ -1575,6 +1575,45 @@ Output:
...
 ```
 
+<!-- TOC --><a name="dataframe_collection"></a>
+### `pydough.dataframe_collection`
+
+The `dataframe_collection` creates a collection within a specified pandas dataframe. This is useful for building datasets dynamically.
+It takes in the following arguments:
+
+- `name`: The name of the dataframe collection.
+- `dataframe`: The panda dataframe containing the corresponding data
```
Comment (hadia206) [inline documentation/dsl.md]:
```suggestion
- `dataframe`: Pandas dataframe containing the corresponding data.
```

File: pydough/user_collections/dataframe_collection.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
+"""
+A user-defined collection of pandas Dataframes.
+Usage:
+`pydough.dataframe_collection(name, dataframe)`
+
+This module defines a collection that generated from a given pandas Dataframe.
+The user must specify the name of the collection and the
+name of the dataframe that will hold the values.
```
Comment (hadia206) [inline pydough/user_collections/dataframe_collection.py]:
```suggestion
name of the dataframe that has the values.
```

File: pydough/user_collections/dataframe_collection.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+The pandas dataframe must have at least one column with one item. The accepted
+types are: NumericType, BooleanType, DatetimeType, StringType and UnknownType.
```
Comment (hadia206) [inline pydough/user_collections/dataframe_collection.py]: This should list the acceptable Pandas Types by the feature not the PyDough ones or you can have both.

File: pydough/user_collections/dataframe_collection.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+Mixed types columsn are not allowed.
```
Comment (hadia206) [inline pydough/user_collections/dataframe_collection.py]: typo.
```suggestion
Mixed types columns are not allowed.
```
Also, does Pandas allow mixed types in a column? My understanding is that will be `object` type. If that's what you're referring to let's clarify that in the docstring as well (i.e. object type is not allowed)

File: pydough/user_collections/dataframe_collection.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+    def __len__(self) -> int:
+        return len(self._dataframe)
+
+    def is_singular(self) -> bool:
+        """Returns True if the collection is guaranteed to contain at most one row."""
+        return len(self) <= 1
+
+    def always_exists(self) -> bool:
+        """Check if the range collection is always non-empty."""
```
Comment (hadia206) [inline pydough/user_collections/dataframe_collection.py]:
```suggestion
        """Check if the dataframe collection is always non-empty."""
```

File: pydough/user_collections/README.md [hunk 6]
```diff
@@ -13,6 +13,12 @@ This module defines the user collections that can be created on the fly and used
     - `end`: The ending value of the range (exclusive).
     - `step`: The step value for incrementing the range. Default is 1.
 
+### [dataframe_collection.py](dataframe_collection.py)
+
+  - `DataframeGeneratedCollection`: Class used to create a dataframe collection using the given dataframe and name.
+    - `name`: The name of the dataframe collection.
+    - `dataframe`: The pandas dataframe containing all data (rows and columns).
```
Comment (hadia206) [inline pydough/user_collections/README.md]:
```suggestion
    - `dataframe`: The Pandas DataFrame containing all data (rows and columns).
```

File: pydough/user_collections/README.md [hunk 7]
```diff
@@ -21,6 +27,11 @@ This module defines the user collections that can be created on the fly and used
     - `end`: The ending value of the range (exclusive).
     - `step`: The step value for incrementing the range. Default is 1.
     - Returns: An instance of `RangeGeneratedCollection`.
+  - `dataframe_collection`: Function to create a dataframe collection with the specified parameters.
+    - `name`: The name of the dataframe collection.
+    - `dataframe`: The pandas dataframe.
```
Comment (hadia206) [inline pydough/user_collections/README.md]:
```suggestion
    - `dataframe`: The Pandas DataFrame.
```

File: pydough/user_collections/user_collection_apis.py [hunk 8]
```diff
@@ -43,3 +46,32 @@ def range_collection(
...
 
     return UnqualifiedGeneratedCollection(range_collection)
+
+
+def dataframe_collection(
+    name: str, dataframe: pd.DataFrame
+) -> UnqualifiedGeneratedCollection:
+    """
+    Implementation of the `pydough.dataframe_collection` function, which provides
+    a way to create a collection of pandas dataframe in PyDough.
```
Comment (hadia206) [inline pydough/user_collections/user_collection_apis.py]:
```suggestion
    a way to create a collection of Pandas DataFrame in PyDough.
```

**Review by hadia206 — COMMENTED** [review]
Good work John. Please see my comment below.

File: pydough/conversion/relational_converter.py [hunk 9]
```diff
@@ -1430,7 +1430,9 @@ def rel_translation(
                 assert context is not None, "Malformed HybridTree pattern."
                 result = self.translate_hybridroot(context)
             case HybridUserGeneratedCollection():
-                assert context is not None, "Malformed HybridTree pattern."
+                if context is None:
+                    context = TranslationOutput(EmptySingleton(), {})
```
[the remark this answers is not in the document]
Comment (knassre-bodo) [reply pydough/conversion/relational_converter.py]: I think this is a change that either I pushed or that I pair-programmed with John, which centers on the edge cases with DataFrame collection testing which somehow flew under the radar with range collections.

The difference here is that in certain situations, the root level of a child subtree can be the generated collection (instead of a normal root node like a collection access or a HybridRoot). In these situations, we need to account for the fact that there is no node above the generated collection, therefore no context for us to join from. The simple fix is to create a sentinel parent to join from (the empty singleton), which is already how we handle other similar things.

In terms of a comment, something like:
`# Account for cases where there is no ancestor preceding the generated collection`

File: documentation/dsl.md [hunk 10]
```diff
@@ -1575,6 +1575,45 @@ Output:
...
+
+Supported Signatures:
+- `dataframe_collection(dataframe, name)`: generates collection with the given datafram and name.
+
+#### Example
```
Comment (knassre-bodo) [inline documentation/dsl.md]: Should we perhaps also have a more involved example that does more things with the collection?

File: pydough/user_collections/dataframe_collection.py [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+        return self._dataframe
+
+    @property
+    def column_names_and_types(self) -> list[tuple[str, PyDoughType]]:
+        assert len(self.columns) == len(self.types)
+        return [(self.columns[i], self.types[i]) for i in range(len(self.columns))]
+
+    @property
+    def unique_column_names(self) -> list[str]:
+        return list(dict.fromkeys(self.columns))
```
Comment (knassre-bodo) [inline pydough/user_collections/dataframe_collection.py]: Aww heck I knew I forgot something in the spec: we need the option to specify subsets of the columns that are unique (just like we do in metadata). It can be an optional argument in `pydough.dataframe_collection` which has type `list[str | list[str]]` where the strings are the column names. For instance, in your `dsl.md` exmaple, it would be `["color", "idx"]` since both columns are unique.

We should then have tests on the behavior of window function / correlation tests with DataFrames that use this argument.

File: tests/test_pipeline_tpch_custom.py [hunk 12]
```diff
@@ -3241,6 +3262,61 @@
             ),
             id="simple_range_5",
         ),
+        pytest.param(
+            PyDoughPandasTest(
+                "r = pydough.range_collection('tbl', 'v', 0, 500, 13).CALCULATE(first_digit=INTEGER(STRING(v)[:1]))\n"
+                "result = r.PARTITION(name='digits', by=first_digit).CALCULATE(first_digit, n=COUNT(tbl))",
+                "TPCH",
```
Comment (knassre-bodo) [inline tests/test_pipeline_tpch_custom.py]: These new tests cover some categories of edge cases John discovered while testing DataFrames which we realized hadn't been covered/handled yet for simple range (hence all the weird HybridTree changes).

File: pydough/user_collections/dataframe_collection.py [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+
+            case _ if pa.types.is_string(field_type) or pa.types.is_large_string(
+                field_type
+            ):
+                return StringType()
+
+            case _ if pa.types.is_binary(field_type) or pa.types.is_large_binary(
+                field_type
+            ):
+                return StringType()
```
Comment (knassre-bodo) [inline pydough/user_collections/dataframe_collection.py]: Shouldn't this be rejected for now?

File: pydough/user_collections/dataframe_collection.py [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+
+            case _ if pa.types.is_struct(field_type):
+                raise ValueError("Structs are not supported for dataframe collections")
+
+            case _ if pa.types.is_dictionary(field_type):
+                raise ValueError(
+                    "Dictionaries are not supported for dataframe collections"
+                )
+            case _:
+                return UnknownType()
```
Comment (knassre-bodo) [inline pydough/user_collections/dataframe_collection.py]: Are there cases where this happens that are valid? If so what are there? (should be documented in comments). Otherwise, we should explicitly ban things with clear error messages.

File: pydough/user_collections/dataframe_collection.py [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+        return self._dataframe
+
+    @property
+    def column_names_and_types(self) -> list[tuple[str, PyDoughType]]:
+        assert len(self.columns) == len(self.types)
+        return list(zip(self.columns, self.types))
+
+    @property
+    def unique_column_names(self) -> list[str]:
+        return list(dict.fromkeys(self.columns))
```
Comment (knassre-bodo) [inline pydough/user_collections/dataframe_collection.py]: Reminder to address the comment I made about about this; this is a serious thing that needs to be added & tested.

File: pydough/user_collections/dataframe_collection.py [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+            f"shape={self.dataframe.shape}, "
+            f"columns={list(self.dataframe.columns)!r}"
+            f")"
+        )
+
+    def equals(self, other) -> bool:
+        return (
+            isinstance(other, DataframeGeneratedCollection)
+            and self.name == other.name
+            and self.columns == other.columns
```
Comment (hadia206) [inline pydough/user_collections/dataframe_collection.py]: Do we need these? If we already check `self.dataframe.equals` in the line below it?

File: documentation/dsl.md [hunk 17]
```diff
@@ -1575,6 +1575,57 @@ Output:
...
+- `name`: The name of the dataframe collection.
+- `dataframe`: Pandas DataFrame containing the corresponding data.
+
+#### Datatypes
+`dataframe_collection` doesn't support mix datatypes in one column.
+
+The supported PyDough types for `dataframe_collection` are:
+- `NumericType`: includes float, integer, infinity, Nan.
+- `BooleanType`: includes classic true and false.
+- `StringType`: includes all aphanumeric caracters.
```
Comment (hadia206) [inline documentation/dsl.md]: Removed and fixed the typo
```suggestion
- `StringType`: alphanumeric characters.
```

File: documentation/dsl.md [hunk 18]
```diff
@@ -1575,6 +1575,57 @@ Output:
...
+- `Datetype`: includes date and datetime.
```
Comment (hadia206) [inline documentation/dsl.md]: Timestamp? NaT?

```suggestion
- `Datetype`: date and datetime.
```

File: pydough/user_collections/dataframe_collection.py [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+        elif pa_types.is_list(field_type) or pa.types.is_large_list(field_type):
+            raise ValueError("Arrays are not supported for dataframe collections")
+
+        elif pa_types.is_struct(field_type):
+            raise ValueError("Structs are not supported for dataframe collections")
+
+        elif pa_types.is_dictionary(field_type):
+            raise ValueError("Dictionaries are not supported for dataframe collections")
+        else:
+            return UnknownType()
```
Comment (hadia206) [inline pydough/user_collections/dataframe_collection.py]: Let's clarify what are these if any are still missing. 
See my comment in the dsl.md as well.

File: pydough/user_collections/dataframe_collection.py (hunk 11 above)
Comment (hadia206) [reply pydough/user_collections/dataframe_collection.py]: Reminder to do this^

File: pydough/user_collections/dataframe_collection.py [hunk 20]
```diff
@@ -0,0 +1,170 @@
...
+                key=lambda t: t.__name__,
+            )
+            if len(types_list) > 1:
+                raise TypeError(
+                    f"Mixed types in column '{col}': {types_list}. All values in a column must be of the same type."
+                )
+            elif len(dataframe[col]) == 0:
+                raise TypeError(
+                    f"Column '{col}' is empty. All columns must have at least one value."
+                )
```
[the remark this answers is not in the document]
Comment (hadia206) [reply pydough/user_collections/dataframe_collection.py]: My understanding from the code is that there's no rows, the code will raise an error. 
@knassre-bodo your test suggestion makes me think that we should allow generating a collection from a DataFrame with no data?

File: pydough/conversion/agg_removal.py [hunk 21]
```diff
@@ -237,7 +237,7 @@ def aggregation_uniqueness_helper(
     match node:
         # Scans are unchanged, and their uniqueness is based on the unique sets
         # of the underlying table.
-        case Scan():
+        case Scan() | GeneratedTable():
```
Comment (hadia206) [inline pydough/conversion/agg_removal.py]: Why moved it here?

File: pydough/user_collections/user_collection_apis.py [hunk 22]
```diff
@@ -49,7 +49,7 @@ def range_collection(
 
 
 def dataframe_collection(
-    name: str, dataframe: pd.DataFrame
+    name: str, dataframe: pd.DataFrame, unique_column_names: list[str | list[str]]
```
Comment (hadia206) [inline pydough/user_collections/user_collection_apis.py]: Should this be optional? Because now it means the user has to specify unique column names which is not intuitive.
Also, I have another request (but that can be a followup if it's too much)
which is having another optional argument specifying some column names to be able create a collection from these columns only.

File: pydough/user_collections/user_collection_apis.py [hunk 23]
```diff
@@ -66,12 +68,26 @@ def dataframe_collection(
         raise TypeError(f"Expected 'name' to be a string, got {type(name).__name__}")
     if not isinstance(dataframe, pd.DataFrame):
         raise TypeError(
-            f"Expected 'dataframe' to be a pandas DataFrame, got {type(dataframe).__name__}"
+            f"Expected 'dataframe' to be a Pandas DataFrame, got {type(dataframe).__name__}"
+        )
+    if not isinstance(unique_column_names, list):
+        raise TypeError(
+            f"Expected 'unique_column_names' to be a list of string, got {type(unique_column_names).__name__}"
```
Comment (hadia206) [inline pydough/user_collections/user_collection_apis.py]: it could also be a list of list of strings

File: documentation/dsl.md [hunk 24]
```diff
@@ -1585,6 +1585,15 @@ It takes in the following arguments:
 
 - `name`: The name of the dataframe collection.
 - `dataframe`: Pandas DataFrame containing the corresponding data.
+- `unique_column_names`: List of strings or list of strings
```
Comment (hadia206) [inline documentation/dsl.md]: This should be `list of strings or list of list of strings`

File: tests/test_pydough_functions/user_collections.py [hunk 25]
```diff
@@ -149,3 +152,776 @@ def user_range_collection_6():
...
+
+    classes_collection = pydough.dataframe_collection(
+        "classes", class_df, ["key", "class_name"]
+    )
+
+    other_classes_same_language = (
+        CROSS(classes_collection.CALCULATE(language, key)).WHERE(
+            (language == original_language) & (key != original_key)
+        )
+        # .CALCULATE(metric=(key != original_key))
```
Comment (knassre-bodo) [inline tests/test_pydough_functions/user_collections.py]:
```suggestion
```

File: tests/test_pipeline_tpch_custom.py [hunk 26]
```diff
@@ -4743,6 +5515,72 @@ def test_pipeline_e2e_tpch_custom(
...
+        ),
+        pytest.param(
+            dataframe_collection_bad_5,
+            None,
+            re.escape(
+                "Arrays in column 'col1', are not supported for dataframe collections"
+            ),
+            id="dataframe_collection_bad_5",
+        ),
+        pytest.param(
```
Comment (knassre-bodo) [inline tests/test_pipeline_tpch_custom.py]: Note: we should have a regular test with a column that is a bad type but is NOT included in the filtered columns, which means it does NOT have an error (e.g. column `x` is an array type, but not included in the filtered columns) since the type checking/inference only matters for the columns we are using.

File: pydough/user_collections/dataframe_collection.py [hunk 27]
```diff
@@ -0,0 +1,291 @@
...
+                f"Fixed-size arrays / tuples in column '{field_name}', are not supported for dataframe collections"
+            )
+
+        else:
+            raise ValueError(
+                f"Unknown type in column '{field_name}' for dataframe collections"
+            )
+
+    @staticmethod
+    def valid_unique_column_names(unique_columns_name: list[str | list[str]]) -> bool:
```
Comment (knassre-bodo) [inline pydough/user_collections/dataframe_collection.py]: We already have something like this in `error_utils.py` used for metadata, called `unique_properties_predicate`. Calling `unique_properties_predicate.verify(unique_columns_names, "unique_columns_names")` verifies that `unique_columns_names` is a non-empty list of objects that are either strings or non-empty lists of strings, and if not raises an exception with the name `"unique_columns_names"` inside the message
